### PR TITLE
NIGH-110 Circle Service Load Config + Logger

### PR DESCRIPTION
**Pull request by alan20854**
Added LoadConfig and Logger for CircleCi Service

### CODE REVIEW

File: internal/clients/diffreviewer/circleci/circleci_service.go [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+	circlelogger "github.com/nightfallai/nightfall_code_scanner/internal/clients/logger/circle_logger"
+	"github.com/nightfallai/nightfall_code_scanner/internal/clients/nightfall"
+	"github.com/nightfallai/nightfall_code_scanner/internal/interfaces/gitdiffintf"
+	"github.com/nightfallai/nightfall_code_scanner/internal/nightfallconfig"
+)
+
+const (
+	WorkspacePathEnvVar        = "GITHUB_WORKSPACE"
+	NightfallAPIKeyEnvVar      = "NIGHTFALL_API_KEY"
+	CircleRepoNameEnvVar       = "CIRCLE_PROJECT_REPONAME"
```
Comment (alan20854) [inline internal/clients/diffreviewer/circleci/circleci_service.go]: may or may not need some of these env vars based on what we need for the github Issues API

File: internal/clients/logger/circle_logger/circle_logger.go [hunk 2]
```diff
@@ -0,0 +1,47 @@
+package circlelogger
```
Comment (bradleymcallister97) [inline internal/clients/logger/circle_logger/circle_logger.go:1]: Can we add tests for these? Should be very similar to the `github_logger_test.go`

File: internal/clients/diffreviewer/circleci/circleci_service.go [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+	CircleOwnerNameEnvVar      = "CIRCLE_PROJECT_USERNAME"
+	CircleCommitShaEnvVar      = "CIRCLE_SHA1"
```
Comment (bradleymcallister97) [inline internal/clients/diffreviewer/circleci/circleci_service.go]: I assume this is the current commit that triggered the action? If so can we rename `CircleCommitShaEnvVar` to be more specific like `CircleCurrentCommitShaEnvVar`

File: internal/clients/diffreviewer/circleci/circleci_service.go [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+	}
+	nightfallConfig, err := nightfallconfig.GetNightfallConfigFile(workspacePath, nightfallConfigFileName)
+	if err != nil {
+		s.Logger.Error("Error getting Nightfall config file. Ensure you have a Nightfall config file located in the root of your repository at .nightfalldlp/config.json with at least one Detector enabled")
+		return nil, err
+	}
+	nightfallAPIKey, ok := os.LookupEnv(NightfallAPIKeyEnvVar)
+	if !ok {
+		s.Logger.Error(fmt.Sprintf("Error getting Nightfall API key. Ensure you have %s set in the Github secrets of the repo", NightfallAPIKeyEnvVar))
+		return nil, errors.New("Missing env var for nightfall api key")
```
Comment (bradleymcallister97) [inline internal/clients/diffreviewer/circleci/circleci_service.go]: nit - I believe we want the errors to start with lower cases (Not sure if we did this for Github Action), but probably good to do it at least for Circle CI. There are also a couple of error messages that start with an upper case above.

**Review by bradleymcallister97 — APPROVED** [review]

File: internal/clients/diffreviewer/circleci/circleci_service.go (hunk 4 above)
Comment (alan20854) [reply internal/clients/diffreviewer/circleci/circleci_service.go]: yup gotcha i'll make sure they all start with lower case

File: internal/clients/diffreviewer/circleci/circleci_service.go [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+		return nil, errors.New("missing env var for commit sha")
+	}
+	beforeCommitSha, ok := os.LookupEnv(CircleBeforeCommitEnvVar)
+	if !ok {
+		s.Logger.Error(fmt.Sprintf("Environment variable %s cannot be found", CircleBeforeCommitEnvVar))
+		return nil, errors.New("missing env var for prev commit sha")
+	}
+	s.GitDiff = &gitdiff.GitDiff{
+		WorkDir:    workspacePath,
+		BaseBranch: "master", //TODO: look into how to get this instead of hardcoding
```
Comment (henkejosh) [inline internal/clients/diffreviewer/circleci/circleci_service.go]: let's not leave TODO's but instead just create a ticket on the board to figure this out

File: internal/clients/diffreviewer/circleci/circleci_service.go [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+// LoadConfig gets all config values from files or environment and creates a config
+func (s *Service) LoadConfig(nightfallConfigFileName string) (*nightfallconfig.Config, error) {
+	s.Logger.Debug("Loading configuration")
+	workspacePath, ok := os.LookupEnv(WorkspacePathEnvVar)
+	if !ok {
+		s.Logger.Error(fmt.Sprintf("Environment variable %s cannot be found", WorkspacePathEnvVar))
+		return nil, errors.New("missing env var for workspace path")
+	}
+	commitSha, ok := os.LookupEnv(CircleCurrentCommitShaEnvVar)
```
Comment (henkejosh) [inline internal/clients/diffreviewer/circleci/circleci_service.go:53]: just checking to make sure - the `LookupEnv` `ok` return value is always true if the variable exists, regardless of whether it's empty. So if any of these Env vars cannot have valid empty values and are also not automatically set by circle, it's probably best to do an additional check to make sure the var isn't empty so we can make sure to log a helpful error message instead of just continuing on (e.g.  of `!ok && var != ""`)

File: internal/clients/diffreviewer/circleci/circleci_service.go [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+		BaseSHA:    beforeCommitSha,
+		Head:       commitSha,
+	}
+	nightfallConfig, err := nightfallconfig.GetNightfallConfigFile(workspacePath, nightfallConfigFileName)
+	if err != nil {
+		s.Logger.Error("Error getting Nightfall config file. Ensure you have a Nightfall config file located in the root of your repository at .nightfalldlp/config.json with at least one Detector enabled")
+		return nil, err
+	}
+	nightfallAPIKey, ok := os.LookupEnv(NightfallAPIKeyEnvVar)
+	if !ok {
```
Comment (henkejosh) [inline internal/clients/diffreviewer/circleci/circleci_service.go]: probably need to add an extra `!ok && nightfallAPIKey != ""`, otherwise if they pass in a NIGHTFALL_API_KEY env var to their configuration but forget to actually set that value in their Circle secrets, this will pass

File: internal/clients/diffreviewer/circleci/circleci_service_test.go [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+		Logger: circleLogger,
+	}
+	return tp
+}
+
+const commitSha = "7b46da6e4d3259b1a1c470ee468e2cb3d9733802"
+const prevCommitSha = "15bf9548d16caff9f398b5aae78a611fc60d55bd"
+const testConfigFileName = "nightfall_test_config.json"
+const excludedCreditCardRegex = "4242-4242-4242-[0-9]{4}"
+const excludedApiToken = "xG0Ct4Wsu3OTcJnE1dFLAQfRgL6b8tIv"
```
Comment (henkejosh) [inline internal/clients/diffreviewer/circleci/circleci_service_test.go:42]: can we add this to the exclusion list in our .nightfalldlp config?

File: internal/clients/diffreviewer/diffutils/diff_filter.go [hunk 9]
```diff
@@ -0,0 +1,46 @@
+package diffutils
+
+import (
+	"strings"
+
+	"github.com/nightfallai/nightfall_code_scanner/internal/clients/diffreviewer"
+)
+
+func FilterFileDiffs(fileDiffs []*diffreviewer.FileDiff) []*diffreviewer.FileDiff {
```
Comment (henkejosh) [inline internal/clients/diffreviewer/diffutils/diff_filter.go:9]: all of these funcs were copy/pasta from github package right? Nothing net new?

Comment (alan20854) [reply internal/clients/diffreviewer/diffutils/diff_filter.go:9]: yup I just pulled them out into their own package

File: internal/clients/diffreviewer/circleci/circleci_service.go [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+// GetLogger gets the github service logger
+func (s *Service) GetLogger() logger.Logger {
+	return s.Logger
+}
+
+// LoadConfig gets all config values from files or environment and creates a config
+func (s *Service) LoadConfig(nightfallConfigFileName string) (*nightfallconfig.Config, error) {
+	s.Logger.Debug("Loading configuration")
+	workspacePath, ok := os.LookupEnv(WorkspacePathEnvVar)
+	if !ok && workspacePath != "" {
```
Comment (henkejosh) [inline internal/clients/diffreviewer/circleci/circleci_service.go]: sorry I was writing quickly - I think we actually want to test that the workspacePath IS empty here (==). Is there any situation in which this can be empty?

File: internal/clients/diffreviewer/circleci/circleci_service.go [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+		s.Logger.Error(fmt.Sprintf("Environment variable %s cannot be found", WorkspacePathEnvVar))
+		return nil, errors.New("missing env var for workspace path")
+	}
+	commitSha, ok := os.LookupEnv(CircleCurrentCommitShaEnvVar)
+	if !ok && commitSha != "" {
```
Comment (henkejosh) [inline internal/clients/diffreviewer/circleci/circleci_service.go]: same ^

File: internal/clients/diffreviewer/circleci/circleci_service.go [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+		s.Logger.Error(fmt.Sprintf("Environment variable %s cannot be found", CircleCurrentCommitShaEnvVar))
+		return nil, errors.New("missing env var for commit sha")
+	}
+	beforeCommitSha, ok := os.LookupEnv(CircleBeforeCommitEnvVar)
+	if !ok && beforeCommitSha != "" {
```
Comment (henkejosh) [inline internal/clients/diffreviewer/circleci/circleci_service.go]: same ^

File: internal/clients/diffreviewer/circleci/circleci_service.go [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+		BaseSHA:    beforeCommitSha,
+		Head:       commitSha,
+	}
+	nightfallConfig, err := nightfallconfig.GetNightfallConfigFile(workspacePath, nightfallConfigFileName)
+	if err != nil {
+		s.Logger.Error("Error getting Nightfall config file. Ensure you have a Nightfall config file located in the root of your repository at .nightfalldlp/config.json with at least one Detector enabled")
+		return nil, err
+	}
+	nightfallAPIKey, ok := os.LookupEnv(NightfallAPIKeyEnvVar)
+	if !ok && nightfallAPIKey != "" {
```
Comment (henkejosh) [inline internal/clients/diffreviewer/circleci/circleci_service.go]: `==`

File: internal/clients/diffreviewer/circleci/circleci_service_test.go [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+}
+
+func (c *circleCiTestSuite) AfterTest(suiteName, testName string) {
+	for _, e := range envVars {
+		err := os.Unsetenv(e)
+		c.NoErrorf(err, "Error unsetting var %s", e)
+	}
+}
+
+func (c *circleCiTestSuite) TestLoadConfig() {
```
Comment (henkejosh) [inline internal/clients/diffreviewer/circleci/circleci_service_test.go:59]: since we have the weirdness w/ the Env variables and checking their existence in the LoadConfig method, lets add a test or two to make sure that we return the correct error if they're empty. We don't need to check all of them but at least the NIGHTFALL_API_KEY

File: internal/clients/diffreviewer/circleci/circleci_service.go (hunk 10 above)
Comment (alan20854) [reply internal/clients/diffreviewer/circleci/circleci_service.go]: shoot my bad haha - and no I don't think so since I'm manually setting it into the environment in the circle orb. I think we should still check that it's not empty however in case it's not there

File: internal/clients/diffreviewer/circleci/circleci_service.go [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+	if !ok || workspacePath == "" {
+		s.Logger.Error(fmt.Sprintf("Environment variable %s cannot be found", WorkspacePathEnvVar))
+		return nil, errors.New("missing env var for workspace path")
+	}
+	commitSha, ok := os.LookupEnv(CircleCurrentCommitShaEnvVar)
+	if !ok || commitSha == "" {
+		s.Logger.Error(fmt.Sprintf("Environment variable %s cannot be found", CircleCurrentCommitShaEnvVar))
+		return nil, errors.New("missing env var for commit sha")
+	}
+	beforeCommitSha, ok := os.LookupEnv(CircleBeforeCommitEnvVar)
```
Comment (henkejosh) [inline internal/clients/diffreviewer/circleci/circleci_service.go]: we don't use this `ok` value, right?

**Review by henkejosh — APPROVED** [review]
